src/Core/Types/Unifier.cs

```diff
@@ -507,7 +507,12 @@ public DataType UnifyStructures(StructureType a, StructureType b)
 				else
 				{
                     var fieldType = Unify(fa.DataType, fb.DataType);
-                    var fieldName = MakeFieldName(fa, fb);
+                    string fieldName;
+                    if (!TryMakeFieldName(fa, fb, out fieldName))
+                        throw new NotSupportedException(
+                            string.Format(
+                                "Failed to unify field '{0}' in structure '{1}' with field '{2}' in structure '{3}'.",
+                                fa.Name, a, fb.Name, b));
```

Comment:
Pleas add a unit test that verifies that this exception is getting thrown correctly.